invenio_records_resources/services/files/transfer.py

```diff
@@ -0,0 +1,156 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2022 CERN.
+#
+# Invenio-Records-Resources is free software; you can redistribute it and/or
+# modify it under the terms of the MIT License; see LICENSE file for more
+# details.
+
+"""Files transfer."""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+
+from invenio_db import db
+from invenio_files_rest.errors import FileSizeError
+
+
+class TransferType(str, Enum):
+    """File type, it inherits from str to be JSON serializable.
+
+    LOCAL represents a file that is stored locally in the instance's storage.
+    EXTERNAL represents a file that needs to be fetched from an external storage
+             and saved locally.
+    LINKED represents a file that is stored externally and is linked to the record.
+    """
+
+    LOCAL = "S"  # S = Stored
+    EXTERNAL = "E"
+    LINKED = "L"
+
+    def __eq__(self, other):
+        """Equality test."""
+        return self.value == other
+
+    def __str__(self):
+        """Return its value."""
+        return self.value
+
+
+class BaseTransfer(ABC):
+    """Local transfer."""
+
+    def __init__(self, type):
+        """Constructor."""
+        self.type = type
+
+    @abstractmethod
+    def init_file(self, record, file_metadata):
+        """Initialize a file."""
+        pass
+
+    def set_file_content(self, record, file, file_key, stream, content_length):
+        """Set file content."""
+        bucket = record.bucket
+        size_limit = bucket.size_limit
+        if content_length and size_limit and content_length > size_limit:
+            desc = (
+                "File size limit exceeded."
+                if isinstance(size_limit, int)
+                else size_limit.reason
+            )
+            raise FileSizeError(description=desc)
```

Comment:
From the point of view of the  user, what would happen if the file is too big? Is this "catchable" so the user can be immediately notified? Without something like messaging I am not sure what would happen since this will run in a task (I suppose?).

Reply:
That is a good point, I will add it to the [error handling issue](https://github.com/inveniosoftware/invenio-records-resources/issues/409). 

On the spot ideas:

a) Notify the user when they make the upload: For this we would need to make the `size` parameter mandatory and trust it's veracity or perform a `HEAD` request.
b) Alternatively, since we would not be able to know the file size in advance we could show the error in the deposit form. How would we persist them?

We can discuss this in depth when tackling the above mentioned issue?